File: csharp/20.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start

using System.Collections.Generic;

public class Solution 
{
    public static Dictionary<char, char> map = new Dictionary<char, char>()
    {
        {'(', ')'},
        {'[', ']'},
        {'{', '}'},
    };

    public bool IsValid(string s) 
    {
        Stack<char> stack = new Stack<char>();

        foreach (var c in s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                stack.Push(c);
            }
            else 
            {
                if (stack.Count == 0)
                {
                    return false;
                }

                var top = stack.Pop();
                if (map[top] != c)
                {
                    return false;
                }
            }
        }

        return stack.Count == 0;
    }
}
// @lc code=end

File: go/20.valid-parentheses.go
func isValid(s string) bool {
	m := map[byte]byte{
		')': '(',
		']': '[',
		'}': '{',
	}
	var stack []byte
	for i := 0; i < len(s); i++ {
		if s[i] == '(' || s[i] == '[' || s[i] == '{' {
			stack = append(stack, s[i])
		} else {
            if len(stack) == 0 {
                return false
            }
			top := stack[len(stack)-1]
			pair := m[s[i]]
			if top == pair {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}